File: src/main/java/net/yakclient/mixin/internal/loader/TargetClassLoader.java
package net.yakclient.mixin.internal.loader;

import net.yakclient.mixin.internal.bytecode.ByteCodeUtils;

import java.io.IOException;


public class TargetClassLoader extends ProxyClassLoader {
    private final PackageTarget target;
//    private final ProxyClassLoader lazyParent;

    public TargetClassLoader(ClassLoader parent, PackageTarget target) {
        super(parent);
//        this.lazyParent = parent;
        this.target = target;
    }

    @Override
    public Class<?> defineClass(String name, byte[] b) {
        return super.defineClass(name, b);
    }

    @Override
    protected final Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
        final ClassTarget target = ClassTarget.create(name);
        if (!this.target.isTargetOf(target)) return super.loadClass(name, resolve);
        try {
            Class<?> c = this.findLoadedClass(name);

            if (c == null) {
                byte[] b = ByteCodeUtils.loadClassBytes(name);
                c = defineClass(name, b, 0, b.length);
            }

            if (resolve) this.resolveClass(c);

            return c;
        } catch (IOException e) {
            throw new ClassNotFoundException(e.getMessage());
        }
    }



    public PackageTarget getTarget() {
        return target;
    }
}

File: base/src/main/kotlin/net/yakclient/mixin/base/internal/loader/TargetClassLoader.kt
package net.yakclient.mixin.base.internal.loader

import net.yakclient.mixin.base.internal.bytecode.ByteCodeUtils.loadClassBytes
import net.yakclient.mixin.base.target.ClassTarget.Companion.of
import net.yakclient.mixin.base.target.PackageTarget
import java.io.IOException

class TargetClassLoader(
    parent: ClassLoader,
    private val target: PackageTarget
) : ClassLoader(parent), ProxyClassLoader {

    @Throws(ClassNotFoundException::class)
    override fun loadClass(name: String, resolve: Boolean): Class<*> {
        val target = of(name)
        return if (!this.target.isTargetOf(target)) super.loadClass(name, resolve) else try {
            var c = findLoadedClass(name)
            if (c == null) {
                val b = loadClassBytes(name)
                c = defineClass(name, b, 0, b!!.size)
            }
            if (resolve) resolveClass(c)
            c
        } catch (e: IOException) {
            throw ClassNotFoundException(e.message)
        }
    }

    override fun defineClass(name: String, b: ByteArray): Class<*> {
       return this.defineClass(name, b, 0, b.size)
    }

    override fun isDefined(cls: String): Boolean {
        return this.findLoadedClass(cls) != null
    }

    override fun loadClass(name: String): Class<*> {
        return super.loadClass(name)
    }
}